From 6053380f8d70e38ef2b0f6dea42191f8fcef2ff8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Micha=C5=82=20G=C3=B3rny?= <mgorny@gentoo.org>
Date: Fri, 16 Feb 2024 11:27:34 +0100
Subject: [PATCH] build: Fix linking to libICE when --enable-libsm is used

Convert `libsm` into a feature, and add libICE as its dependency,
in order to fix underlinking.  The code in `libxfce4ui/xfce-sm-client.c`
is using the functions from `<X11/ICE/ICElib.h>` when `--enable-libsm`
is used.  If libxfce4ui is compiled with a more strict linker (this
could be simulated e.g. by passing `-Wl,-z,defs` to GNU ld), it fails
due to missing function references:

```
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: .libs/libxfce4ui_2_la-xfce-sm-client.o: in function `xsmp_new_ice_connection':
/tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:840:(.text+0xa80): undefined reference to `IceConnectionNumber'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:839:(.text+0xa99): undefined reference to `IceConnectionNumber'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:842:(.text+0xab5): undefined reference to `IceConnectionNumber'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: .libs/libxfce4ui_2_la-xfce-sm-client.o: in function `xsmp_ice_init':
/tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:875:(.text+0x220a): undefined reference to `IceSetIOErrorHandler'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:876:(.text+0x221d): undefined reference to `IceSetIOErrorHandler'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:881:(.text+0x2239): undefined reference to `IceAddConnectionWatch'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: .libs/libxfce4ui_2_la-xfce-sm-client.o: in function `xsmp_process_ice_messages':
/tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:806:(.text+0x2875): undefined reference to `IceProcessMessages'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:810:(.text+0x28c5): undefined reference to `IceSetShutdownNegotiation'
/usr/lib/gcc/x86_64-pc-linux-gnu/13/../../../../x86_64-pc-linux-gnu/bin/ld: /tmp/libxfce4ui/libxfce4ui/xfce-sm-client.c:814:(.text+0x28ec): undefined reference to `IceCloseConnection'
```

Originally reported to https://bugs.gentoo.org/924321
---
 configure.ac                | 11 ++++++++--
 libxfce4ui/Makefile.am      |  2 ++
 libxfce4ui/xfce-sm-client.c | 42 ++++++++++++++++++-------------------
 3 files changed, 32 insertions(+), 23 deletions(-)

diff --git a/configure.ac b/configure.ac
index 806ec3b5..cdfc27d5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -19,6 +19,7 @@ m4_define([libxfce4util_min_version], [4.17.2])
 m4_define([xfconf_min_version], [4.12.0])
 
 m4_define([libx11_min_version], [1.6.7])
+m4_define([libice_min_version], [1.0.10])
 m4_define([libsm_min_version], [1.2.3])
 
 m4_define([libstartup_notif_min_version], [0.4])
@@ -159,7 +160,13 @@ if test x"$ENABLE_X11" != x"yes"; then
   enable_libsm=no
   enable_startup_notification=no
 fi
-XDT_CHECK_OPTIONAL_PACKAGE([LIBSM], [sm], [libsm_min_version], [libsm], [session management library])
+XDT_CHECK_OPTIONAL_FEATURE([LIBSM],
+                           [libsm],
+                           [
+                             XDT_FEATURE_DEPENDENCY([LIBICE], [ice], [libice_min_version])
+                             XDT_FEATURE_DEPENDENCY([LIBSM], [sm], [libsm_min_version])
+                           ],
+                           [session management library])
 XDT_CHECK_OPTIONAL_PACKAGE([LIBSTARTUP_NOTIFICATION],
                            [libstartup-notification-1.0],
                            [libstartup_notif_min_version], [startup-notification],
@@ -418,7 +425,7 @@ echo "* Keyboard library support:  no"
 fi
 echo "* X11 Support:               ${ENABLE_X11:-no}"
 if test x"$ENABLE_X11" = x"yes"; then
-echo "  * Libsm support:           ${LIBSM_FOUND:-no}"
+echo "  * Libsm support:           ${ENABLE_LIBSM:-no}"
 echo "  * Startup notif support:   ${LIBSTARTUP_NOTIFICATION_FOUND:-no}"
 fi
 echo "* Wayland Support:           ${ENABLE_WAYLAND:-no}"
diff --git a/libxfce4ui/Makefile.am b/libxfce4ui/Makefile.am
index fe58c153..9777c851 100644
--- a/libxfce4ui/Makefile.am
+++ b/libxfce4ui/Makefile.am
@@ -82,11 +82,13 @@ libxfce4ui_sources += \
 	xfce-sm-client.h
 
 libxfce4ui_2_la_CFLAGS += \
+	$(LIBICE_CFLAGS) \
 	$(LIBSM_CFLAGS) \
 	$(LIBX11_CFLAGS) \
 	$(LIBSTARTUP_NOTIFICATION_CFLAGS)
 
 libxfce4ui_2_la_LIBADD += \
+	$(LIBICE_LIBS) \
 	$(LIBSM_LIBS) \
 	$(LIBX11_LIBS) \
 	$(LIBSTARTUP_NOTIFICATION_LIBS)
diff --git a/libxfce4ui/xfce-sm-client.c b/libxfce4ui/xfce-sm-client.c
index decccf6e..17751d3d 100644
--- a/libxfce4ui/xfce-sm-client.c
+++ b/libxfce4ui/xfce-sm-client.c
@@ -51,7 +51,7 @@
 #include <unistd.h>
 #endif
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
 #include <X11/ICE/ICElib.h>
 #include <X11/SM/SMlib.h>
 #endif
@@ -138,7 +138,7 @@ struct _XfceSMClient
 {
     GObject parent;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     SmcConn session_connection;
     IceConn ice_connection;
 #endif
@@ -230,7 +230,7 @@ static void xfce_sm_client_finalize(GObject *obj);
 static void xfce_sm_client_set_client_id(XfceSMClient *sm_client,
                                          const gchar *client_id);
 static void xfce_sm_client_parse_argv(XfceSMClient *sm_client);
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
 static void xfce_sm_client_set_property_from_command(XfceSMClient *sm_client,
                                                      const char *property_name,
                                                      gchar **command,
@@ -588,7 +588,7 @@ xfce_sm_client_finalize(GObject *obj)
     startup_options.client_id = NULL;
     startup_options.sm_disable = FALSE;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(sm_client->session_connection)
         xfce_sm_client_disconnect(sm_client);
 #endif
@@ -657,7 +657,7 @@ xfce_sm_client_set_clone_command(XfceSMClient *sm_client,
     g_return_if_fail(XFCE_IS_SM_CLIENT(sm_client));
     sm_client->clone_command = copy_command(sm_client->clone_command,
                                             clone_command);
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     xfce_sm_client_set_property_from_command(sm_client, SmCloneCommand,
                                              sm_client->clone_command,
                                              SM_ARG_REMOVE);
@@ -740,7 +740,7 @@ G_GNUC_END_IGNORE_DEPRECATIONS
     sm_client->argv = NULL;
 }
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
 
 static void
 xfce_sm_client_set_state(XfceSMClient *sm_client,
@@ -1567,7 +1567,7 @@ gboolean
 xfce_sm_client_connect(XfceSMClient *sm_client,
                        GError **error)
 {
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     char buf[256] = "";
     unsigned long mask;
     SmcCallbacks callbacks;
@@ -1585,7 +1585,7 @@ xfce_sm_client_connect(XfceSMClient *sm_client,
     if(startup_options.sm_disable)
         return TRUE;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     xsmp_ice_init(sm_client);
 
     mask = SmcSaveYourselfProcMask | SmcDieProcMask | SmcSaveCompleteProcMask
@@ -1736,7 +1736,7 @@ xfce_sm_client_disconnect(XfceSMClient *sm_client)
     if(startup_options.sm_disable)
         return;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(G_UNLIKELY(!sm_client->session_connection)) {
         g_warning("%s() called with no session connection", G_STRFUNC);
         return;
@@ -1766,7 +1766,7 @@ gboolean
 xfce_sm_client_is_connected(XfceSMClient *sm_client)
 {
     g_return_val_if_fail(XFCE_IS_SM_CLIENT(sm_client), FALSE);
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     return !!sm_client->session_connection;
 #else
     return FALSE;
@@ -1892,7 +1892,7 @@ xfce_sm_client_set_desktop_file(XfceSMClient *sm_client,
         }
     }
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(sm_client->session_connection) {
         SmProp prop, *props[1];
         SmPropValue propval;
@@ -1907,7 +1907,7 @@ xfce_sm_client_set_desktop_file(XfceSMClient *sm_client,
 
         SmcSetProperties(sm_client->session_connection, 1, props);
     }
-#endif /* HAVE_LIBSM */
+#endif /* ENABLE_LIBSM */
 
 out:
     if(rcfile)
@@ -1941,7 +1941,7 @@ xfce_sm_client_request_shutdown(XfceSMClient *sm_client,
 
     /* TODO: support xfce4-session's DBus interface */
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(G_LIKELY(sm_client->session_connection)) {
         SmcRequestSaveYourself(sm_client->session_connection, SmSaveBoth,
                                True, SmInteractStyleAny, False, True);
@@ -1967,7 +1967,7 @@ xfce_sm_client_set_restart_style(XfceSMClient *sm_client,
 
     sm_client->restart_style = restart_style;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(sm_client->session_connection) {
         SmProp prop, *props[1];
         SmPropValue propval;
@@ -1985,7 +1985,7 @@ xfce_sm_client_set_restart_style(XfceSMClient *sm_client,
 
         SmcSetProperties(sm_client->session_connection, 1, props);
     }
-#endif /* HAVE_LIBSM */
+#endif /* ENABLE_LIBSM */
 
     g_object_notify(G_OBJECT(sm_client), "restart-style");
 }
@@ -2010,7 +2010,7 @@ xfce_sm_client_set_priority(XfceSMClient *sm_client,
 
     sm_client->priority = priority;
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(sm_client->session_connection) {
         SmProp prop, *props[1];
         SmPropValue propval;
@@ -2025,7 +2025,7 @@ xfce_sm_client_set_priority(XfceSMClient *sm_client,
 
         SmcSetProperties(sm_client->session_connection, 1, props);
     }
-#endif /* HAVE_LIBSM */
+#endif /* ENABLE_LIBSM */
 
     g_object_notify(G_OBJECT(sm_client), "priority");
 }
@@ -2052,7 +2052,7 @@ xfce_sm_client_set_current_directory(XfceSMClient *sm_client,
     g_free(sm_client->current_directory);
     sm_client->current_directory = g_strdup(current_directory);
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(sm_client->session_connection) {
         SmProp prop, *props[1];
         SmPropValue propval;
@@ -2067,7 +2067,7 @@ xfce_sm_client_set_current_directory(XfceSMClient *sm_client,
 
         SmcSetProperties(sm_client->session_connection, 1, props);
     }
-#endif /* HAVE_LIBSM */
+#endif /* ENABLE_LIBSM */
 
     g_object_notify(G_OBJECT(sm_client), "current-directory");
 }
@@ -2107,7 +2107,7 @@ xfce_sm_client_set_restart_command(XfceSMClient *sm_client,
     g_return_if_fail(XFCE_IS_SM_CLIENT(sm_client));
     sm_client->restart_command = copy_command(sm_client->restart_command,
                                               restart_command);
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     xfce_sm_client_set_property_from_command(sm_client, SmRestartCommand,
                                              sm_client->restart_command,
                                              SM_ARG_APPEND);
@@ -2231,7 +2231,7 @@ xfce_sm_client_get_state_file(XfceSMClient *sm_client)
 
     g_free(resource);
 
-#ifdef HAVE_LIBSM
+#ifdef ENABLE_LIBSM
     if(G_LIKELY(sm_client->state_file) && sm_client->session_connection)
     {
         gchar *discard_command[4];
-- 
GitLab

